core/src/main/scala/kafka/log/Log.scala

```diff
@@ -798,9 +800,26 @@ class Log(@volatile var dir: File,
           }
         } else {
           // we are taking the offsets we are given
-          if (!appendInfo.offsetsMonotonic || appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset)
-            throw new IllegalArgumentException(s"Out of order offsets found in append to $topicPartition: " +
-              records.records.asScala.map(_.offset))
+          if (!appendInfo.offsetsMonotonic)
+            throw new OffsetsOutOfOrderException(s"Out of order offsets found in append to $topicPartition: " +
+                                                 records.records.asScala.map(_.offset))
+
+          if (appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset) {
```

Comment:
This is an existing issue. I am wondering if we can tighten up the check here a bit. Currently, if records are of pre v2 format and have multiple batches, appendInfo.firstOrLastOffset returns the last offset of the last batch. In this case, it's better to use the last offset of the first batch for this check.